File: src/Physics_Model.jl
#**************************************************************************************
# Physics_Model.jl
# ================ part of the GeoEfficiency.jl package.
#
# here is the place where all physical elements is being modeled and created as computer objects.
#
#**************************************************************************************

import Base: show, isless


#--------------Point---------------------------------------------

"""

    Point(Height::Real, Rho::Real)

construct and return a `Point` source that can be a source of itself or an
`anchor point` of a source.

 *  `Height` : point height relative to the detector.

 *  `Rho` : point off axis relative to the detector axis of symmetry.

# Note please

Each detector type give interpretation to the height in a different way as follow:

 *  for `CylDetector` the point source `height` is consider to be measured from the detector `face surface`.

 *  for `BoreDetector` the point source `height` is consider to be measured from the `detector middle`, +ve value are above the detector center while -ve are below.

 *  for `WellDetector` the point source `height` is considered to be measured from the detector `hole surface`.
 
"""
mutable struct Point
	Height::Float64
	Rho::Float64
	Point(Height::Real, Rho::Real) = new(float(Height), float(Rho))
end #type

"""

	Point(Height::Real)

the same as `Point(Height::Real, Rho::Real)` but return an axial point.

`Height` : point height relative to the detector. \n

"""
Point(Height::Real) = Point(Height, 0.0)

"""

	Point()

the same as `Point(Height::Real, Rho::Real)` but ask the user to provide the
information from the `console`. 

"""
function Point()
	print_with_color(:yellow,"\n II- The Radiactive Source Anchor Point:-\n")
	Height = getfloat("\n\t > Height (cm) = ")
	Rho = getfloat("\n\t > Off-axis (cm) = ")
	Point(Height, Rho)
end #function

function setHeight!(aPnt::Point, xH::Real)
	aPnt.Height = xH
	return aPnt
end #function
function setHeight(aPnt::Point, xH::Real)
	pnt::Point = deepcopy(aPnt)
	pnt.Height = xH
	return pnt
end #function
function setRho!(aPnt::Point, xRho::Real)
	aPnt.Rho = xRho
	return aPnt
end #function
function setRho(aPnt::Point, xRho::Real)
	pnt::Point = deepcopy(aPnt)
	pnt.Rho = xRho
	return pnt
end #function
id(aPnt::Point) = "Point[Height=$(aPnt.Height), Rho=$(aPnt.Rho)]"
show(pnt::Point) = print(id(pnt))


#--------------source---------------------------------------------

"""

	source(anchorPnt::Point = Point()

return a tuple describing the source (`anchorPnt`, `SrcRadius`, `SrcLength`) based on the user input to the `console`.

 *  `aPnt` : the source anchoring point. if it is missing the used is asked to input it via the `console`.

 *  `SrcRadius` : source radius.

 *  `SrcLength` : source length.

If the global `GeoEfficiency_isPoint` is set to true both `SrcRadius` and `SrcLength` are set to zero.

""" 
function source(anchorPnt::Point = Point())
    
	setSrcToPoint() && return (anchorPnt, 0.0, 0.0)

	SrcRadius = getfloat("\n\t > Source Radius (cm) = ")
    if 0.0 != SrcRadius
        SrcLength = getfloat("\n\t > Source Length (cm) = ")
		println()
		warn("Only axial non-point is allaowed Now: the off-axis will be set to Zero")
        anchorPnt.Rho = 0.0

	else
        SrcLength = 0.0

	end #if
    return (anchorPnt, SrcRadius, SrcLength)
end #function


#----------------Detector------------------------------------


abstract type RadiationDetector end
show(io::IO, detector::RadiationDetector) = print(id(detector))

"abstract base of all the gamma detectors"
RadiationDetector
const Detector = RadiationDetector

isless(detector1::RadiationDetector, detector2::RadiationDetector) = isless(volume(detector1), volume(detector2))


#--------------CylDetector----------------------------------------

"""

	CylDetector(CryRadius::Real, CryLength::Real)

return a `cylindrical` detector of the given crystal dimmensions.

 *  `CryRadius` : the detector crystal radius.

 *  `CryLength` : the detector crystal length.
 
"""
struct CylDetector <: RadiationDetector
	CryRadius::Float64    	#Real
    CryLength::Float64		#Real

	function CylDetector(CryRadius::Real, CryLength::Real)
		@assert Inf > CryRadius > 0.0	  	"Crystal Radius: expect +ve number, get $(CryRadius)."
		@assert Inf > CryLength >= 0.0  	"Crystal Length: expect +ve number or zero, get) $(CryLength)."
		new(float(CryRadius), float(CryLength))
	end #if

end #type

"""

    CylDetector(CryRadius::Real)

return a cylindrical detector with crystal length 0.0.

`CryRadius` : the detector crystal radius.

"""
CylDetector(CryRadius::Real) = CylDetector(CryRadius, 0.0)

"""

    CylDetector()

return a cylindrical detector according to the input from the `console`.

"""
function CylDetector()
	print_with_color(:yellow," I- The Cylinderical Detetcor physical Dimensions:-\n")
	CryRadius = getfloat("\n\t > Crystal Radius (cm) = ")
	CryLength = getfloat("\n\t > Crystal Length (cm) = ")
	CylDetector(CryRadius, CryLength)
end #function

id(detector::CylDetector) = "CylDetector[CryRadius=$(detector.CryRadius), CryLength=$(detector.CryLength)]"
volume(detector::CylDetector) = pi * detector.CryRadius^2 * detector.CryLength 


#-------------BoreDetector-------------------------------------------


"""

	BoreDetector(CryRadius::Real, CryLength::Real, HoleRadius::Real)

return a `bore-hole` detector.

 *  `CryRadius` : the detector crystal radius.

 *  `CryLength` : the detector crystal length.

 *  `HoleRadius` : the detector hole radius.
 
"""
struct BoreDetector <: RadiationDetector
	CryRadius::Float64    	#Real
    CryLength::Float64    	#Real
	HoleRadius::Float64    	#Real

	function BoreDetector(CryRadius::Real, CryLength::Real, HoleRadius::Real)
		@assert Inf > CryRadius > 0.0		"Crystal Radius: expect +ve number, get $(CryRadius)."
		@assert Inf > CryLength > 0.0		"Crystal Length: expect +ve number, get $(CryLength)."
		@assert CryRadius > HoleRadius > 0.0	"Hole Radius: expect +ve number Less than 'Crystal Radius=$(CryRadius)', get $(HoleRadius)."
		new(float(CryRadius), float(CryLength), float(HoleRadius))
	end #if

end #type

"""

	BoreDetector()

return a bore-hole detector according to the input from the `console`.

"""
function BoreDetector()
	print_with_color(:yellow," I- The Bore Hole Detetcor physical Dimensions:-\n")
	CryRadius  = getfloat("\n\t > Crystal Radius (cm) = ")
	CryLength  = getfloat("\n\t > Crystal Length (cm) = ")
	HoleRadius = getfloat("\n\t > Hole Radius (cm) = ", 0.0, CryRadius)
	BoreDetector(CryRadius, CryLength, HoleRadius)
end #function

id(detector::BoreDetector) = "BoreDetector[CryRadius=$(detector.CryRadius), CryLength=$(detector.CryLength), HoleRadius=$(detector.HoleRadius)]"
volume(detector::BoreDetector) = pi * (detector.CryRadius^2 - detector.HoleRadius ^2 )* detector.CryLength 


#-----------WellDetector------------------------------------------------------

"""

	WellDetector(CryRadius::Real, CryLength::Real, HoleRadius::Real, HoleDepth::Real)

return a Well-Type detector.

 *  `CryRadius` : the detector crystal radius.

 *  `CryLength` : the detector crystal length.

 *  `HoleRadius` : the detector hole radius.

 *  `HoleDepth` : the detector hole length.
 
"""
struct WellDetector <: RadiationDetector
	CryRadius::Float64
    CryLength::Float64
	HoleRadius::Float64
	HoleDepth::Float64

	function WellDetector(CryRadius::Real, CryLength::Real, HoleRadius::Real, HoleDepth::Real)
		@assert Inf > CryRadius > 0.0				"Crystal Radius: expect +ve number, get $(CryRadius)."
		@assert Inf > CryLength > 0.0				"Crystal Length: expect +ve number, get $(CryLength)."
		@assert Inf > CryRadius > HoleRadius > 0.0	"Hole Radius: expect +ve number Less than 'Crystal Radius=$(CryRadius)', get $(HoleRadius)."
		@assert CryLength > HoleDepth > 0.0	   	"Hole Depth: expect +ve number Less than 'Crystal Length=$(CryLength)', get $(HoleDepth)."
		new(float(CryRadius), float(CryLength), float(HoleRadius), float(HoleDepth))
	end #if

end #type

"""

	WellDetector()

return a Well-Type detector according to the input from the `console`.

"""
function WellDetector()
	print_with_color(:yellow," I- The Well-Type Detetcor physical Dimensions:-\n")
	CryRadius  = getfloat("\n\t > Crystal Radius (cm) = ")
	CryLength  = getfloat("\n\t > Crystal Length (cm) = ")
	HoleRadius = getfloat("\n\t > Hole Radius (cm) = ", 0.0, CryRadius)
	HoleDepth  = getfloat("\n\t > Hole Radius (cm) = ", 0.0, CryLength)
	WellDetector(CryRadius, CryLength, HoleRadius, HoleDepth)
end #function

id(detector::WellDetector) = "WellDetector[CryRadius=$(detector.CryRadius), CryLength=$(detector.CryLength), HoleRadius=$(detector.HoleRadius), HoleDepth=$(detector.HoleDepth)]"
volume(detector::WellDetector) = pi * (detector.CryRadius^2 * detector.CryLength - detector.HoleRadius ^2 * detector.HoleDepth)


#----------RadiationDetector-------------------------------------------------------

"""

	Detector()

construct and return an object of the RadiationDetector type (`CylDetector`, `BoreDetector` or `WellDetector`)
according to the input from the console.

# Note please

* this methode aquire all required information from the `console` and will prompt user on invalid data.

* if any method with argument(s) take an `invalid` argument it would throw an error.

* if the value the last argument is `zero` of a method with `more` than one argument it behave as a missing argument.

"""
function RadiationDetector()
	print_with_color(:yellow, "\n I- The detector physical Dimensions :-\n")
	CryRadius  = getfloat("\n\t > Crystal Radius (cm) = ")
	CryLength  = getfloat("\n\t > Crystal Length (cm) = ")
	HoleRadius = getfloat("\n(zero for cylindrical detectors) > Hole Radius (cm) = ", 0.0, CryRadius)
	if   0.0 == HoleRadius
		return CylDetector(CryRadius, CryLength)

	else
		HoleDepth = getfloat("\n(zero for Bore-Hole detectors) > Hole Depth (cm) = ", 0.0, CryLength )
		return 0.0 == HoleDepth ?
				BoreDetector(CryRadius, CryLength, HoleRadius) :
				WellDetector(CryRadius, CryLength, HoleRadius, HoleDepth)
	end #if
end #function

"""

	Detector(CryRadius::Real)

return cylindrical(or Disk) detector with `CryLength` equal to zero.

"""
RadiationDetector(CryRadius::Real) = CylDetector(CryRadius)

"""

	Detector(CryRadius::Real, CryLength::Real)

return cylindrical detector.

"""
RadiationDetector(CryRadius::Real, CryLength::Real) = CylDetector(CryRadius, CryLength)

"""

	Detector(CryRadius::Real, CryLength::Real, HoleRadius::Real)

return bore-hole or cylindrical detector if `HoleRadius` = 0.0

"""
RadiationDetector(CryRadius::Real, CryLength::Real, HoleRadius::Real) = 0.0 == HoleRadius ?
				                                        CylDetector(CryRadius, CryLength) :
				                                        BoreDetector(CryRadius, CryLength, HoleRadius)

"""

	Detector(CryRadius::Real, CryLength::Real, HoleRadius::Real, HoleDepth::Real)

return well-type or bore-hole or cylindrical detector according to the arguments.

"""
RadiationDetector(CryRadius::Real, CryLength::Real, HoleRadius::Real, HoleDepth::Real) = 0.0 == HoleDepth ?
				                                        RadiationDetector(CryRadius, CryLength, HoleRadius):
				                                        WellDetector(CryRadius, CryLength, HoleRadius, HoleDepth)

"""

	Detector(detector::RadiationDetector)

return just the inputed detector

"""
RadiationDetector(detector::RadiationDetector) = detector

File: GeoEfficiency_Python/geoefficiency/physics_model.py
"""
Physics model classes for GeoEfficiency package.

This module contains classes representing physical objects:
- Point: A point in space (source or anchor point)
- Detector: Base detector class
- CylDetector: Cylindrical detector
- BoreDetector: Bore-hole (coaxial) detector
- WellDetector: Well-type detector
"""

import math
from .error import validate_detector, invalid_geometry, not_implemented_error


# ============================================================================
# Point Class
# ============================================================================

class Point:
    """
    Represents a point in space relative to a detector.
    
    A Point can represent either a point source or an anchor point for
    higher-dimension sources (disc, cylinder, etc.).
    
    Parameters
    ----------
    height : float
        Point height relative to the detector (interpretation varies by detector type)
    rho : float, optional
        Point off-axis distance from detector axis of symmetry (default: 0.0)
    
    Attributes
    ----------
    height : float
        Point height
    rho : float
        Off-axis distance
    
    Notes
    -----
    Height interpretation by detector type:
    - CylDetector: measured from detector face surface
    - BoreDetector: measured from detector middle (+ve above, -ve below)
    - WellDetector: measured from detector hole surface
    """
    
    def __init__(self, height, rho=0.0):
        self.height = float(height)
        self.rho = float(rho)
    
    def __repr__(self):
        return f"Point(height={self.height}, rho={self.rho})"
    
    def __str__(self):
        return f"Point[Height={self.height}, Rho={self.rho}]"
    
    def __eq__(self, other):
        if not isinstance(other, Point):
            return False
        return math.isclose(self.height, other.height) and math.isclose(self.rho, other.rho)
    
    def __lt__(self, other):
        """Compare points by height, then by rho."""
        if not isinstance(other, Point):
            return NotImplemented
        if not math.isclose(self.height, other.height):
            return self.height < other.height
        return self.rho < other.rho


# ============================================================================
# Detector Classes
# ============================================================================

class Detector:
    """
    Base class for radiation detectors.
    
    This abstract base class defines the interface for detector types.
    """
    
    def __init__(self):
        pass
    
    def __repr__(self):
        return f"{self.__class__.__name__}()"


class CylDetector(Detector):
    """
    Cylindrical detector.
    
    A cylindrical detector with specified crystal radius and length.
    
    Parameters
    ----------
    cry_radius : float
        Crystal (active) radius in cm
    cry_length : float
        Crystal (active) length (height) in cm
    
    Attributes
    ----------
    cry_radius : float
        Crystal radius
    cry_length : float
        Crystal length
    
    Raises
    ------
    InValidDetectorDim
        If cry_radius or cry_length is not positive
    """
    
    def __init__(self, cry_radius, cry_length):
        super().__init__()
        cry_radius = float(cry_radius)
        cry_length = float(cry_length)
        
        validate_detector(cry_radius > 0, f"Crystal radius must be positive, got {cry_radius}")
        validate_detector(cry_length > 0, f"Crystal length must be positive, got {cry_length}")
        
        self.cry_radius = cry_radius
        self.cry_length = cry_length
    
    def __repr__(self):
        return f"CylDetector(cry_radius={self.cry_radius}, cry_length={self.cry_length})"
    
    def __str__(self):
        return f"Cylindrical Detector[Radius={self.cry_radius}cm, Length={self.cry_length}cm]"
    
    def __eq__(self, other):
        if not isinstance(other, CylDetector):
            return False
        return (math.isclose(self.cry_radius, other.cry_radius) and
                math.isclose(self.cry_length, other.cry_length))


class BoreDetector(Detector):
    """
    Bore-hole (coaxial) detector.
    
    A cylindrical detector with a hole drilled through its center.
    
    Parameters
    ----------
    cry_radius : float
        Outer crystal radius in cm
    cry_length : float
        Crystal length in cm
    hole_radius : float
        Hole radius in cm
    
    Attributes
    ----------
    cry_radius : float
        Outer crystal radius
    cry_length : float
        Crystal length
    hole_radius : float
        Hole radius
    
    Raises
    ------
    InValidDetectorDim
        If dimensions are invalid (hole must be smaller than crystal radius)
    """
    
    def __init__(self, cry_radius, cry_length, hole_radius):
        super().__init__()
        cry_radius = float(cry_radius)
        cry_length = float(cry_length)
        hole_radius = float(hole_radius)
        
        validate_detector(cry_radius > 0, f"Crystal radius must be positive, got {cry_radius}")
        validate_detector(cry_length > 0, f"Crystal length must be positive, got {cry_length}")
        validate_detector(hole_radius >= 0, f"Hole radius must be non-negative, got {hole_radius}")
        validate_detector(hole_radius < cry_radius, 
                         f"Hole radius {hole_radius} must be less than crystal radius {cry_radius}")
        
        self.cry_radius = cry_radius
        self.cry_length = cry_length
        self.hole_radius = hole_radius
    
    def __repr__(self):
        return (f"BoreDetector(cry_radius={self.cry_radius}, cry_length={self.cry_length}, "
                f"hole_radius={self.hole_radius})")
    
    def __str__(self):
        return (f"Bore-hole Detector[Outer Radius={self.cry_radius}cm, Length={self.cry_length}cm, "
                f"Hole Radius={self.hole_radius}cm]")
    
    def __eq__(self, other):
        if not isinstance(other, BoreDetector):
            return False
        return (math.isclose(self.cry_radius, other.cry_radius) and
                math.isclose(self.cry_length, other.cry_length) and
                math.isclose(self.hole_radius, other.hole_radius))


class WellDetector(Detector):
    """
    Well-type detector.
    
    A cylindrical detector with a well (hole) for inserting samples.
    
    Parameters
    ----------
    cry_radius : float
        Outer crystal radius in cm
    cry_length : float
        Crystal length in cm
    hole_radius : float
        Well radius in cm
    hole_depth : float
        Well depth in cm
    
    Attributes
    ----------
    cry_radius : float
        Outer crystal radius
    cry_length : float
        Crystal length
    hole_radius : float
        Well radius
    hole_depth : float
        Well depth
    
    Raises
    ------
    InValidDetectorDim
        If dimensions are invalid
    """
    
    def __init__(self, cry_radius, cry_length, hole_radius, hole_depth):
        super().__init__()
        cry_radius = float(cry_radius)
        cry_length = float(cry_length)
        hole_radius = float(hole_radius)
        hole_depth = float(hole_depth)
        
        validate_detector(cry_radius > 0, f"Crystal radius must be positive, got {cry_radius}")
        validate_detector(cry_length > 0, f"Crystal length must be positive, got {cry_length}")
        validate_detector(hole_radius > 0, f"Hole radius must be positive, got {hole_radius}")
        validate_detector(hole_depth > 0, f"Hole depth must be positive, got {hole_depth}")
        validate_detector(hole_radius < cry_radius,
                         f"Hole radius {hole_radius} must be less than crystal radius {cry_radius}")
        validate_detector(hole_depth <= cry_length,
                         f"Hole depth {hole_depth} must not exceed crystal length {cry_length}")
        
        self.cry_radius = cry_radius
        self.cry_length = cry_length
        self.hole_radius = hole_radius
        self.hole_depth = hole_depth
    
    def __repr__(self):
        return (f"WellDetector(cry_radius={self.cry_radius}, cry_length={self.cry_length}, "
                f"hole_radius={self.hole_radius}, hole_depth={self.hole_depth})")
    
    def __str__(self):
        return (f"Well Detector[Radius={self.cry_radius}cm, Length={self.cry_length}cm, "
                f"Well Radius={self.hole_radius}cm, Well Depth={self.hole_depth}cm]")
    
    def __eq__(self, other):
        if not isinstance(other, WellDetector):
            return False
        return (math.isclose(self.cry_radius, other.cry_radius) and
                math.isclose(self.cry_length, other.cry_length) and
                math.isclose(self.hole_radius, other.hole_radius) and
                math.isclose(self.hole_depth, other.hole_depth))


# ============================================================================
# Source Type Management
# ============================================================================

# Global source type tracking
_src_type = None
_SRC_UNKNOWN = 0
_SRC_POINT = 1
_SRC_DISC = 2
_SRC_CYLINDER = 3


def typeofSrc():
    """
    Get the current source type.
    
    Returns
    -------
    int
        Current source type code
    """
    return _src_type if _src_type is not None else _SRC_UNKNOWN


def setSrcToPoint(value=True):
    """
    Set source type to point source.
    
    Parameters
    ----------
    value : bool, optional
        If True, set to point source; if False, unset (default: True)
    
    Returns
    -------
    bool
        True if source was set to point, False otherwise
    """
    global _src_type
    if value:
        _src_type = _SRC_POINT
        return True
    else:
        _src_type = _SRC_UNKNOWN
        return False


def source(anchor_pnt=None):
    """
    Get source information (anchor point, radius, length).
    
    Parameters
    ----------
    anchor_pnt : Point, optional
        Source anchor point. If None, creates a point at (0, 0)
    
    Returns
    -------
    tuple
        (anchor_point, src_radius, src_length)
    
    Notes
    -----
    For point sources, both radius and length are set to zero.
    """
    if anchor_pnt is None:
        anchor_pnt = Point(0.0, 0.0)
    
    if typeofSrc() == _SRC_POINT:
        return (anchor_pnt, 0.0, 0.0)
    else:
        # Non-point source handling would go here
        # For now, return as point source
        return (anchor_pnt, 0.0, 0.0)
